### PR TITLE
feat(data): Add CBOR support for V2 API

### LINKED ISSUE #2894: [V2] Implement CBOR support
# 🚀 Feature Request

### Relevant Package

 This feature request is for core-data

### Description
 In current V2 API implementation, we haven't implement the binary support.

### Describe the solution you'd like
 Implement the proper CBOR support in core-data

### PR DESCRIPTION
**Pull request by judehung**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/master/.github/Contributing.md.

## Issue Number:
fix #2894 

## What is the new behavior?
Add implementation for decoding CBOR-encoded AddEventRequest

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by judehung** [comment]
rebased

**Comment by judehung** [comment]
recheck

File: internal/core/data/v2/controller/http/event.go
Comment (lenny-intel) [inline internal/core/data/v2/controller/http/event.go:69]: This implementation takes in a CBOR request and then publishes the request to the MessageBus as JSON. We are expecting it to publish the original CBOR to the MessageBus. V1, does this asynchronously. 
@cloudxxx8 , @jpwhitemn We should decide if V2 should also publish the CBOR asynchronous to  validation and persisting.

File: internal/core/data/v2/io/event.go
Comment (lenny-intel) [inline internal/core/data/v2/io/event.go:65]: The DTO should be implementing `UnmarshalCBOR` just like it does for JSON. The cbor module the same corresponding `Unmarshaler interface` which is called from `cbor.Unmarshal`
```
type Unmarshaler interface {
	UnmarshalCBOR([]byte) error
}
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/core/data/v2/controller/http/event.go
Comment (judehung) [reply internal/core/data/v2/controller/http/event.go:69]: @lenny-intel , I have added another commit, https://github.com/edgexfoundry/edgex-go/pull/3202/commits/a9e89425e353be16965c3e2b4be2414c6f565315, to update func PublishEvent to publish CBOR request to MessageBus in CBOR encoding byte array.  We could then open another issue for asynchronous PublishEvent once the community makes the final decision.

File: internal/core/data/v2/io/event.go
Comment (judehung) [reply internal/core/data/v2/io/event.go:65]: @lenny-intel I've submitted a PR at https://github.com/edgexfoundry/go-mod-core-contracts/pull/527 to implement UnmarshalCBOR fot AddEventRequest DTO.  Once the PR is merged, I will update the go.mod and then remove the validation code here.

File: internal/core/data/v2/application/event.go
Comment (lenny-intel) [inline internal/core/data/v2/application/event.go:99]: Problem here  with CBOR (and actually with JSON also) is we are re-encoding the request DTO. If no error occurred, we know the request received is good so is there a way we can just pass PublishEvent() the encoded data?

I recommend that `ReadAddEventRequest` also return the encoded []byte data for the request that can then be passed to PublishEvent(). 

@jpwhitemn,  Then the question for Core WG is do we want the Publish to be async like it is in V1 for CBOR, i.e. don't wait for validation and persistence. Just get it onto the MessageBus ASAP.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: internal/core/data/v2/application/event.go
Comment (judehung) [reply internal/core/data/v2/application/event.go:99]: @lenny-intel, re-encoding the request DTO is a necessary step here for both CBOR and JSON as we set ApiVerion on the request DTO prior to publishing to MessageBus.  Your recommendation will work if we move the code of setting ApiVersion to `ReadAddEventRequest`.

Comment (lenny-intel) [reply internal/core/data/v2/application/event.go:99]: @judehung , `ApiVersion` is now required in the request DTO , so it will not pass validation if not already present. Also, it is already set in the factory method. We should be able to Publish the original encoded data if no errors occurred.

Since the app services can handle the request directly from Device Services, the request is also validated, so we could publish async to Core data processing. @jpwhitemn we should discuss this in Core WG.

Comment (jpwhitemn) [reply internal/core/data/v2/application/event.go:99]: Per Core WG - DECISION: publish initially encoded object to (not re-encoding) async to message bus (app service) while simultaneously validate/persist in (separate go function)
If persistence fails, it does not impact message publication (cow is already out of the barn)
Let app service (or whatever is on the other end of the message bus) deal with bad/invalid data

**Comment by judehung** [comment]
rebased and add one more commits to address the latest review comment at https://github.com/edgexfoundry/edgex-go/pull/3202#discussion_r587618347

**Comment by judehung** [comment]
recheck

**Comment by judehung** [comment]
This PR is hitting a data race issue from go-mod-messaging below.  This PR should be suspended until the data race issue is resolved.
```
WARNING: DATA RACE

[2021-03-05T08:59:34.772Z] Read at 0x00c0003035b0 by goroutine 11:

[2021-03-05T08:59:34.772Z]   github.com/edgexfoundry/go-mod-messaging/v2/internal/pkg/zeromq.(*zeromqClient).Publish()

[2021-03-05T08:59:34.772Z]       /go/pkg/mod/github.com/edgexfoundry/go-mod-messaging/v2@v2.0.0-dev.3/internal/pkg/zeromq/client.go:69 +0x5b

[2021-03-05T08:59:34.772Z]   github.com/edgexfoundry/edgex-go/internal/core/data/v2/application.PublishEvent()

[2021-03-05T08:59:34.772Z]       /w/workspace/edgex-go/11/internal/core/data/v2/application/event.go:83 +0x664

[2021-03-05T08:59:34.772Z] 

[2021-03-05T08:59:34.772Z] Previous write at 0x00c0003035b0 by goroutine 9:

[2021-03-05T08:59:34.772Z]   github.com/edgexfoundry/go-mod-messaging/v2/internal/pkg/zeromq.(*zeromqClient).bindToPort()

[2021-03-05T08:59:34.772Z]       /go/pkg/mod/github.com/edgexfoundry/go-mod-messaging/v2@v2.0.0-dev.3/internal/pkg/zeromq/client.go:238 +0x139

[2021-03-05T08:59:34.772Z]   github.com/edgexfoundry/go-mod-messaging/v2/internal/pkg/zeromq.(*zeromqClient).Publish()

[2021-03-05T08:59:34.772Z]       /go/pkg/mod/github.com/edgexfoundry/go-mod-messaging/v2@v2.0.0-dev.3/internal/pkg/zeromq/client.go:70 +0x2bb

[2021-03-05T08:59:34.772Z]   github.com/edgexfoundry/edgex-go/internal/core/data/v2/application.PublishEvent()

[2021-03-05T08:59:34.772Z]       /w/workspace/edgex-go/11/internal/core/data/v2/application/event.go:83 +0x664
```

**Comment by lenny-intel** [comment]
> This PR is hitting a data race issue as shown below.

Issue created: https://github.com/edgexfoundry/go-mod-messaging/issues/95

**Comment by lenny-intel** [comment]
@judehung , race condition has been addressed in go-mod-messaging. Please update to latest version and verify this did in fact resolve the race condition you saw here. THX!

**Comment by judehung** [comment]
rebased and update go.mod with the usage of github.com/edgexfoundry/go-mod-messaging/v2 v2.0.0-dev.6 to avoid the data race issue as mentioned in https://github.com/edgexfoundry/go-mod-messaging/issues/95

**Comment by judehung** [comment]
> @judehung , race condition has been addressed in go-mod-messaging. Please update to latest version and verify this did in fact resolve the race condition you saw here. THX!

@lenny-intel after upgrading to latest go-mod-messaging, this PR is no longer hitting the data race.  Thanks!

File: internal/core/data/v2/application/event.go
Comment (lenny-intel) [inline internal/core/data/v2/application/event.go:79]: Please add the full topic the event is published to, i.e. `publishTopic` below.

File: internal/core/data/v2/io/event.go
Comment (lenny-intel) [inline internal/core/data/v2/io/event.go:23]: Do we need to make this a service configuration setting? i.e. some systems may need this to be lower while others may handle it being larger?

If so please add a TODO here and an issue to make it a setting.

File: internal/core/data/v2/io/event_test.go
Comment (lenny-intel) [inline internal/core/data/v2/io/event_test.go:29]: Use factory methods

File: internal/core/data/v2/io/event_test.go
Comment (lenny-intel) [inline internal/core/data/v2/io/event_test.go:41]: Use factory methods

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Comment by judehung** [comment]
rebased and push one more commit for review comments

File: internal/core/data/v2/io/event_test.go
Comment (judehung) [reply internal/core/data/v2/io/event_test.go]: updated https://github.com/edgexfoundry/edgex-go/pull/3202/commits/a97e7d088cfea97e2f7e544ceb56995fb0b64fad

File: internal/core/data/v2/io/event_test.go
Comment (judehung) [reply internal/core/data/v2/io/event_test.go]: updated https://github.com/edgexfoundry/edgex-go/pull/3202/commits/a97e7d088cfea97e2f7e544ceb56995fb0b64fad

File: internal/core/data/v2/io/event.go
Comment (judehung) [reply internal/core/data/v2/io/event.go]: updated https://github.com/edgexfoundry/edgex-go/pull/3202/commits/a97e7d088cfea97e2f7e544ceb56995fb0b64fad

File: internal/core/data/v2/application/event.go
Comment (judehung) [reply internal/core/data/v2/application/event.go]: updated https://github.com/edgexfoundry/edgex-go/pull/3202/commits/a97e7d088cfea97e2f7e544ceb56995fb0b64fad

**Review by lenny-intel — APPROVED** [review]
LGTM

**Pull request merged**
